### DISCUSSION TITLE
json-schema validation of the input document

### DISCUSSION THREAD

**Discussion by ewout8** [open-policy-agent/community#327] [OPA and Rego] [answered]
Hi all,

I would like to suggest the possibility of specifying json-schema validation of the input document in the OPA server configuration file.

Currently I have two policies in separate packages. Both contain checks on the input document: is the necessary input present, is it in the right format and so on. It would be great if I could specify a json-schema file and resolve-directory for each package in the OPA server configuration file. When getting data that requires input the OPA server will first validate the input provided against the configured json-schema file, resolving any references using the configured resolve-directory. If validation fails, the OPA server returns status code 400. If validation is succesful it's business as usual.

**Comment by anderseknert** [comment, marked as answer]
Hi @ewout8 👋 

This is currently WIP: https://github.com/open-policy-agent/opa/pull/5572

**Reply by ewout8** [reply]
Hi @anderseknert ,

wow, that's quick!

Great that this is WIP already! If I understand correctly it will not be as flexible as my suggested implementation. But I could work around that by running separate OPA servers for separate input schemas and supplying a schema file with all references already resolved. Thanks for the info!

**Reply by anderseknert** [reply]
I don't think you'll need to run separate OPA servers in order to have different schemas, but perhaps @ashutosh-narkar can elaborate on the implementation.

**Reply by ashutosh-narkar** [reply]
@ewout8 the idea is you can provide a single schema file or a directory that contains multiple schemas and OPA would perform input validation. Does that work for your case? If you could share any example inputs and schemas that would be great!

**Reply by ewout8** [reply]
@ashutosh-narkar unfortunately I'm not allowed to share work related content, so I can only describe my OPA setup. I have an OPA server running with 3 loaded packages. Package A en B implement different poliices, package C contains general rules used by A and B. The rules in A and B have different input requirements, which are checked by extra input validation rules in each package. It would be great if I could replace the extra input validation rules in A and B by specifying a json-schema file for each package in the server's configuration file. My json-schema's for input specification contain references to other json-schema's. Currently all json-schema files are in one directory, but as the number of policies, managing parties and json-schema files grows there will be a need to use different directories for json-schema management. So it would also be great if I could specify a resolve-directory for each package in the server's configuration file as well.

Hope this clarifies what I am looking for.

**Reply by anderseknert** [reply]
Rule-specific schemas should definitely be an option. As for directories, I'm thinking that (filesystem) paradigm might be the wrong one for runtime validation. Rather than providing `--schema` and/or `--input-validation` as command line flags to the server, we should allow both schemas and preferences for runtime validation to be included in bundles along with policies and data.
